In [12]:
import opendatasets as od
#od.download('stackoverflow-developer-survey-2020')


0it [00:00, ?it/s]

100%|█████████▉| 94560256/94603888 [01:44<00:00, 910853.22it/s]

In [13]:
import os

In [14]:
os.listdir('stackoverflow-developer-survey-2020')    # verify that the dataset was downloaded into the directory stackoverflow-developer-survey-2020

['README.txt', 'survey_results_public.csv', 'survey_results_schema.csv']

In [15]:
import pandas as pd


16384it [00:01, 14497.60it/s]           
8192it [00:00, 12822.26it/s]            
94609408it [01:45, 895485.68it/s]                              


In [16]:
survey_raw_df = pd.read_csv('stackoverflow-developer-survey-2020/survey_results_public.csv')


In [17]:
survey_raw_df

Respondent                                         MainBranch Hobbyist  \
0               1                     I am a developer by profession      Yes   
1               2                     I am a developer by profession       No   
2               3                        I code primarily as a hobby      Yes   
3               4                     I am a developer by profession      Yes   
4               5  I used to be a developer by profession, but no...      Yes   
...           ...                                                ...      ...   
64456       64858                                                NaN      Yes   
64457       64867                                                NaN      Yes   
64458       64898                                                NaN      Yes   
64459       64925                                                NaN      Yes   
64460       65112                                                NaN      Yes   

        Age Age1stCode CompFreq  CompTotal  ConvertedComp             Country  \
0       NaN         13  Monthly        NaN            NaN             Germany   
1       NaN         19      NaN        NaN            NaN      United Kingdom   
2       NaN         15      NaN        NaN            NaN  Russian Federation   
3      25.0         18      NaN        NaN            NaN             Albania   
4      31.0         16      NaN        NaN            NaN       United States   
...     ...        ...      ...        ...            ...                 ...   
64456   NaN         16      NaN        NaN            NaN       United States   
64457   NaN        NaN      NaN        NaN            NaN             Morocco   
64458   NaN        NaN      NaN        NaN            NaN            Viet Nam   
64459   NaN        NaN      NaN        NaN            NaN              Poland   
64460   NaN        NaN      NaN        NaN            NaN               Spain   

         CurrencyDesc  ...                  SurveyEase           SurveyLength  \
0       European Euro  ...  Neither easy nor difficult  Appropriate in length   
1      Pound sterling  ...                         NaN                    NaN   
2                 NaN  ...  Neither easy nor difficult  Appropriate in length   
3        Albanian lek  ...                         NaN                    NaN   
4                 NaN  ...                        Easy              Too short   
...               ...  ...                         ...                    ...   
64456             NaN  ...                         NaN                    NaN   
64457             NaN  ...                         NaN                    NaN   
64458             NaN  ...                         NaN                    NaN   
64459             NaN  ...                         NaN                    NaN   
64460             NaN  ...                         NaN                    NaN   

      Trans                                     UndergradMajor  \
0        No  Computer science, computer engineering, or sof...   
1       NaN  Computer science, computer engineering, or sof...   
2       NaN                                                NaN   
3        No  Computer science, computer engineering, or sof...   
4        No  Computer science, computer engineering, or sof...   
...     ...                                                ...   
64456   NaN  Computer science, computer engineering, or sof...   
64457   NaN                                                NaN   
64458   NaN                                                NaN   
64459   NaN                                                NaN   
64460   NaN  Computer science, computer engineering, or sof...   

            WebframeDesireNextYear                      WebframeWorkedWith  \
0                     ASP.NET Core                    ASP.NET;ASP.NET Core   
1                              NaN                                     NaN   
2                              NaN                           

In [18]:
survey_raw_df.columns

Index(['Respondent', 'MainBranch', 'Hobbyist', 'Age', 'Age1stCode', 'CompFreq',
       'CompTotal', 'ConvertedComp', 'Country', 'CurrencyDesc',
       'CurrencySymbol', 'DatabaseDesireNextYear', 'DatabaseWorkedWith',
       'DevType', 'EdLevel', 'Employment', 'Ethnicity', 'Gender', 'JobFactors',
       'JobSat', 'JobSeek', 'LanguageDesireNextYear', 'LanguageWorkedWith',
       'MiscTechDesireNextYear', 'MiscTechWorkedWith',
       'NEWCollabToolsDesireNextYear', 'NEWCollabToolsWorkedWith', 'NEWDevOps',
       'NEWDevOpsImpt', 'NEWEdImpt', 'NEWJobHunt', 'NEWJobHuntResearch',
       'NEWLearn', 'NEWOffTopic', 'NEWOnboardGood', 'NEWOtherComms',
       'NEWOvertime', 'NEWPurchaseResearch', 'NEWPurpleLink', 'NEWSOSites',
       'NEWStuck', 'OpSys', 'OrgSize', 'PlatformDesireNextYear',
       'PlatformWorkedWith', 'PurchaseWhat', 'Sexuality', 'SOAccount',
       'SOComm', 'SOPartFreq', 'SOVisitFreq', 'SurveyEase', 'SurveyLength',
       'Trans', 'UndergradMajor', 'WebframeDesireNextYear',
  

The columns are coded. the questions are in another file: survey_results_schema.csv 

In [19]:
schema_fname = 'stackoverflow-developer-survey-2020/survey_results_schema.csv'
schema_raw = pd.read_csv(schema_fname, index_col='Column').QuestionText    #load it as Pandas Series with Column as the index and the QuestionText as the value

In [20]:
schema_raw


Column
Respondent            Randomized respondent ID number (not in order ...
MainBranch            Which of the following options best describes ...
Hobbyist                                        Do you code as a hobby?
Age                   What is your age (in years)? If you prefer not...
Age1stCode            At what age did you write your first line of c...
                                            ...                        
WebframeWorkedWith    Which web frameworks have you done extensive d...
WelcomeChange         Compared to last year, how welcome do you feel...
WorkWeekHrs           On average, how many hours per week do you wor...
YearsCode             Including any education, how many years have y...
YearsCodePro          NOT including education, how many years have y...
Name: QuestionText, Length: 61, dtype: object

In [21]:
schema_raw['YearsCodePro']

'NOT including education, how many years have you coded professionally (as a part of your work)?'

# preprocessing & cleaning the data for our analysis.

# Data Preparation & Cleaning

While the survey responses contain a wealth of information, we'll limit our analysis to the following areas:

* Demographics of the survey respondents and the global programming community
* Distribution of programming skills, experience, and preferences
* Employment-related information, preferences, and opinions

Let's select a subset of columns with the relevant data for our analysis.

In [22]:
selected_columns = [
    # Demographics
    'Country',
    'Age',
    'Gender',
    'EdLevel',
    'UndergradMajor',
    # Programming experience
    'Hobbyist',
    'Age1stCode',
    'YearsCode',
    'YearsCodePro',
    'LanguageWorkedWith',
    'LanguageDesireNextYear',
    'NEWLearn',
    'NEWStuck',
    # Employment
    'Employment',
    'DevType',
    'WorkWeekHrs',
    'JobSat',
    'JobFactors',
    'NEWOvertime',
    'NEWEdImpt'
]

In [23]:
len(selected_columns)

20

Let's extract a copy of the data from these columns into a new data frame survey_df. We can continue to modify further without affecting the original data frame.

In [24]:
survey_df = survey_raw_df[selected_columns].copy()

In [25]:
schema = schema_raw[selected_columns]

In [26]:
survey_df.shape

(64461, 20)

In [27]:
survey_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64461 entries, 0 to 64460
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Country                 64072 non-null  object 
 1   Age                     45446 non-null  float64
 2   Gender                  50557 non-null  object 
 3   EdLevel                 57431 non-null  object 
 4   UndergradMajor          50995 non-null  object 
 5   Hobbyist                64416 non-null  object 
 6   Age1stCode              57900 non-null  object 
 7   YearsCode               57684 non-null  object 
 8   YearsCodePro            46349 non-null  object 
 9   LanguageWorkedWith      57378 non-null  object 
 10  LanguageDesireNextYear  54113 non-null  object 
 11  NEWLearn                56156 non-null  object 
 12  NEWStuck                54983 non-null  object 
 13  Employment              63854 non-null  object 
 14  DevType                 49370 non-null

### convert some other columns into numeric data types while ignoring any non-numeric value. The non-numeric are converted to NaN

In [28]:
survey_df['Age1stCode'] = pd.to_numeric(survey_df.Age1stCode, errors='coerce')
survey_df['YearsCode'] = pd.to_numeric(survey_df.YearsCode, errors='coerce')
survey_df['YearsCodePro'] = pd.to_numeric(survey_df.YearsCodePro, errors='coerce')

In [29]:
# view some basic statistics about numeric columns
survey_df.describe()

Age    Age1stCode     YearsCode  YearsCodePro   WorkWeekHrs
count  45446.000000  57473.000000  56784.000000  44133.000000  41151.000000
mean      30.834111     15.476572     12.782051      8.869667     40.782174
std        9.585392      5.114081      9.490657      7.759961     17.816383
min        1.000000      5.000000      1.000000      1.000000      1.000000
25%       24.000000     12.000000      6.000000      3.000000     40.000000
50%       29.000000     15.000000     10.000000      6.000000     40.000000
75%       35.000000     18.000000     17.000000     12.000000     44.000000
max      279.000000     85.000000     50.000000     50.000000    475.000000

There seems to be a problem with the age column, as the minimum value is 1 and the maximum is 279

 ignore the rows where the age is higher than 100 years or lower than 10 years as invalid survey responses. 
 

In [30]:
survey_df.drop(survey_df[survey_df.Age < 10].index, inplace=True)
survey_df.drop(survey_df[survey_df.Age > 100].index, inplace=True)

The same holds for WorkWeekHrs. Let's ignore entries where the value for the column is higher than 140 hours. (~20 hours per day).

In [31]:
survey_df.drop(survey_df[survey_df.WorkWeekHrs > 140].index, inplace=True)

The gender column also allows for picking multiple options. We'll remove values containing more than one option to simplify our analysis.

In [32]:
survey_df['Gender'].value_counts()

Man                                                            45895
Woman                                                           3835
Non-binary, genderqueer, or gender non-conforming                385
Man;Non-binary, genderqueer, or gender non-conforming            121
Woman;Non-binary, genderqueer, or gender non-conforming           92
Woman;Man                                                         73
Woman;Man;Non-binary, genderqueer, or gender non-conforming       25
Name: Gender, dtype: int64

In [33]:
import numpy as np

In [34]:
survey_df.where(~(survey_df.Gender.str.contains(';', na=False)), np.nan, inplace=True)

In [35]:
survey_df['Gender'].value_counts()

Man                                                  45895
Woman                                                 3835
Non-binary, genderqueer, or gender non-conforming      385
Name: Gender, dtype: int64

In [36]:
survey_df.sample(10)

Country   Age Gender  \
8923   Republic of Korea   NaN    Man   
55874            Romania  31.0    Man   
18300            Ukraine   NaN    Man   
37618                NaN   NaN    NaN   
24889            Germany  23.0    Man   
45271             Poland  21.0    Man   
7639               India   NaN    NaN   
63250      United States   NaN    NaN   
58957      United States  34.0  Woman   
17386             Canada   NaN    Man   

                                               EdLevel  \
8923      Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
55874     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
18300     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
37618                                              NaN   
24889  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
45271     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
7639   Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
63250                                              NaN   
58957     Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
17386       Other doctoral degree (Ph.D., Ed.D., etc.)   

                                          UndergradMajor Hobbyist  Age1stCode  \
8923   A business discipline (such as accounting, fin...      Yes        25.0   
55874  Computer science, computer engineering, or sof...       No        15.0   
18300  A humanities discipline (such as literature, h...      Yes         NaN   
37618                                                NaN      Yes         NaN   
24889  Computer science, computer engineering, or sof...      Yes        14.0   
45271  Computer science, computer engineering, or sof...      Yes        17.0   
7639   Computer science, computer engineering, or sof...       No        19.0   
63250                                                NaN      Yes         NaN   
58957  A humanities discipline (such as literature, h...      Yes        21.0   
17386  Computer science, computer engineering, or sof...      Yes        16.0   

       YearsCode  YearsCodePro  \
8923         4.0           2.0   
55874       11.0           8.0   
18300        5.0           5.0   
37618        NaN           NaN   
24889        8.0           4.0   
45271        4.0           1.0   
7639         8.0           3.0   
63250        NaN           NaN   
58957        8.0           8.0   
17386       34.0          25.0   

                                      LanguageWorkedWith  \
8923                                  Python;R;Scala;SQL   
55874                                           Java;SQL   
18300                              JavaScript;TypeScript   
37618                                                NaN   
24889  C++;Go;Haskell;HTML/CSS;Java;JavaScript;Kotlin...   
45271  Assembly;C;C#;C++;Java;JavaScript;Kotlin;PHP;P...   
7639        C;C#;C++;HTML/CSS;Java;JavaScript;TypeScript   
63250                Bash/Shell/PowerShell;C#;Python;SQL   
58957                     HTML/CSS;JavaScript;TypeScript   
17386                                                NaN   

                                  LanguageDesireNextYear  \
8923                                          C++;Kotlin   
55874                                                NaN   
18300                                 Haskell;JavaScript   
37618                                                NaN   
24889  Go;HTML/CSS;Java;JavaScript;Kotlin;R;SQL;TypeS...   
45271                                Dart;Go;Scala;Swift   
7639                                              Python   
63250                  Bash/Shell/PowerShell;C#;Java;SQL   
58957                     HTML/CSS;JavaScript;TypeScript   
17386                    C++;HTML/CSS;JavaScript;PHP;SQL   

                   NEWLearn  \
8923            Once a year   
55874  Once every few years   
18300      Every few months   
37618                   NaN   
24889      Every few months   
45271           Once a year   
7639       Every few months   
63250  Once every few years   
58957         Once a decade   
17386  Once every few years